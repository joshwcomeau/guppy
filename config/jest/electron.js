// @flow
global.window = {
  require: require,
};
