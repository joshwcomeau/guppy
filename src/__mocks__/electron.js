/* eslint-disable flowtype/require-valid-file-annotation */
import * as path from 'path';

module.exports = {
  ipcRenderer: {
    send: jest.fn(),
  },
  remote: {
    app: {
      getAppPath: () => path.resolve(__dirname, '..', '..', '..'),
      getPath: () =>
        process.env.APPDATA ||
        (process.platform === 'darwin'
          ? process.env.HOME + 'Library/Preferences'
          : '/var/local'),
    },
    dialog: {
      showErrorBox: jest.fn(),
      showOpenDialog: jest.fn(),
      showMessageBox: jest.fn(),
    },
    shell: {
      moveItemToTrash: jest.fn(),
    },
  },
};
