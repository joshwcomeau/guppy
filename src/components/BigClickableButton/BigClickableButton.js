// @flow
import React, { Component } from 'react';
import { Spring, animated } from 'react-spring';
import styled from 'styled-components';
import Color from 'color';

import { COLORS, GRADIENTS } from '../../constants';

type Props = {
  width: number,
  height: number,
  thickness: number,
  colors: Array<string>,
  children?: React$Node,
  isOn: boolean,
  onClick: (ev: SyntheticMouseEvent<*>) => void,
};

type State = {
  isActive: boolean,
};

class BigClickableButton extends Component<Props, State> {
  static defaultProps = {
    width: 64,
    height: 64,
    thickness: 6,
    colors: GRADIENTS.darkPrimary,
  };

  state = {
    status: 'off',
    isActive: false,
  };

  depress = () => this.setState({ isActive: true });
  release = () => this.setState({ isActive: false });

  render() {
    const {
      width,
      height,
      thickness,
      colors,
      onClick,
      isOn,
      children,
    } = this.props;
    const { isActive } = this.state;

    const bottomColor = Color(colors[0])
      .darken(0.3)
      .rgb()
      .string();
    const sideColor = Color(colors[1])
      .darken(0.3)
      .rgb()
      .string();

    const background = `linear-gradient(45deg, ${colors.join(', ')})`;

    // When the user is actively clicking on it, it should be all-the-way
    // depressed (full thickness).
    // Otherwise, the amount of depression will depend on whether or not the
    // button is "on".
    // transition order is always 0 => thickness =>  thickness - 2
    const amountDepressedOptions = [0, thickness, thickness - 2];

    const depressedToIndex = isActive ? 1 : isOn ? 2 : 0;
    const depressedFromIndex =
      depressedToIndex === 0 ? 2 : depressedToIndex - 1;

    return (
      <OuterWrapper thickness={thickness}>
        <Spring
          from={{ offset: amountDepressedOptions[depressedFromIndex] }}
          to={{ offset: amountDepressedOptions[depressedToIndex] }}
          config={{
            tension: 196,
            friction: 16,
          }}
          native
        >
          {interpolated => (
            <Wrapper width={width} height={height} offset={interpolated.offset}>
              <Button
                width={width}
                height={height}
                background={background}
                onMouseDown={this.depress}
                onMouseUp={this.release}
                onMouseLeave={this.release}
                onClick={onClick}
              >
                {children}
              </Button>
              <ButtonSideWrapper>
                <SidesSvg
                  style={{ display: 'block' }}
                  width={width + thickness + 2}
                  height={height + thickness + 2}
                >
                  <path
                    d={`
                      M 0,${height}
                      L ${width},${height}
                      L ${width + thickness + 2},${height + thickness + 2}
                      L ${thickness},${height + thickness + 2}
                    `}
                    fill={bottomColor}
                  />
                  <path
                    d={`
                      M ${width},0
                      L ${width + thickness + 2},${thickness}
                      L ${width + thickness + 2},${height + thickness + 2}
                      L ${width}, ${height}
                    `}
                    fill={sideColor}
                  />
                </SidesSvg>
              </ButtonSideWrapper>
            </Wrapper>
          )}
        </Spring>
      </OuterWrapper>
    );
  }
}

const OuterWrapper = styled.div`
  position: relative;
  overflow: hidden;
  padding-right: ${props => props.thickness}px;
  padding-bottom: ${props => props.thickness}px;
`;

export const Wrapper = animated(styled.div`
  position: relative;
  width: ${props => props.width}px;
  height: ${props => props.height}px;
  transform: ${props => `translate(${props.offset}px, ${props.offset}px)`};
`);

export const Button = styled.button`
  width: ${props => props.width}px;
  height: ${props => props.height}px;
  display: flex;
  align-items: center;
  justify-content: space-around;
  color: ${COLORS.textOnBackground};
  font-size: 18px;

  border: none;
  background: ${props => props.background};
  outline: none; /* TODO: Better a11y story :/ */
`;

export const ButtonSideWrapper = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  pointer-events: none;
`;

const SidesSvg = styled.svg``;

export default BigClickableButton;
