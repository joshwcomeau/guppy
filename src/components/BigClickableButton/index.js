// @flow
export { default } from './BigClickableButton';
