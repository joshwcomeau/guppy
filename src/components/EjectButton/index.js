// @flow
export { default } from './EjectButton';
