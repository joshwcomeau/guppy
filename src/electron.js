/**
 * This is our main Electron process.
 * It handles opening our app window, and quitting the application.
 */
const { app, BrowserWindow, ipcMain, dialog } = require('electron');
const { autoUpdater } = require('electron-updater');
const log = require('electron-log');
const path = require('path');
const url = require('url');
const fixPath = require('fix-path');
const chalkRaw = require('chalk');

const killProcessId = require('./services/kill-process-id.service');
const electronStore = require('./services/electron-store.service');

const chalk = new chalkRaw.constructor({ level: 3 });
let icon256Path = '../public/256x256.png';

// In production, we need to use `fixPath` to let Guppy use NPM.
// For reasons unknown, the opposite is true in development; adding this breaks
// everything.
if (process.env.NODE_ENV !== 'development') {
  icon256Path = '../build/256x256.png';
  fixPath();
}

// Keep a global reference of the window object, if you don't, the window will
// be closed automatically when the JavaScript object is garbage collected.
let mainWindow;

// Guppy is primarily a process runner, and so many processes may be started.
// To ensure that Guppy cleans up after itself, we'll store all of the spawned
// process IDs up here in the main process, so that we can terminate them all
// before the app quits.
let processIds = [];

// Used for renderer tasks on close (emitting event app-will-close once)
let emitAppWillClose = true;

const MOVE_TO_APP_FOLDER_KEY = 'leave-application-in-original-location';

// This logging setup is not required for auto-updates to work,
// but it sure makes debugging easier :)
//-------------------------------------------------------------------
autoUpdater.logger = log;
autoUpdater.logger.transports.file.level = 'info';

// default logs from electron-updater are
// 'Checking for update', 'Found verison ...', 'Downloading update ...'

autoUpdater.on('update-downloaded', (event, releaseNotes, releaseName) => {
  const dialogOpts = {
    type: 'info',
    buttons: ['Restart', 'Later'],
    title: 'Application Update',
    message: process.platform === 'win32' ? releaseNotes : releaseName,
    detail:
      'A new version has been downloaded. Restart the application to apply the updates.',
  };

  dialog.showMessageBox(dialogOpts, response => {
    if (response === 0) autoUpdater.quitAndInstall();
  });
});

function createWindow() {
  // Check if there is a newer version
  autoUpdater.checkForUpdatesAndNotify();

  // Verifies if Guppy is already in the Applications folder
  // and prompts the user to move it if it isn't
  manageApplicationLocation();

  // Create the browser window.
  mainWindow = new BrowserWindow({
    width: 1120,
    height: 768,
    minWidth: 777,
    titleBarStyle: 'hidden',
    icon: path.join(__dirname, icon256Path),
  });

  // set up some chrome extensions
  if (process.env.NODE_ENV === 'development') {
    const {
      default: installExtension,
      REACT_DEVELOPER_TOOLS,
      REACT_PERF,
    } = require('electron-devtools-installer');

    require('electron-debug')({ showDevTools: true });

    const ChromeLens = {
      // ID of the extension (https://chrome.google.com/webstore/detail/chromelens/idikgljglpfilbhaboonnpnnincjhjkd)
      id: 'idikgljglpfilbhaboonnpnnincjhjkd',
      electron: '>=1.2.1',
    };

    const extensions = [REACT_DEVELOPER_TOOLS, REACT_PERF, ChromeLens];

    for (const extension of extensions) {
      try {
        installExtension(extension);
      } catch (e) {
        console.error(chalk.red(`[ELECTRON] Extension installation failed`), e);
      }
    }
  }

  // and load the index.html of the app.
  const startUrl =
    process.env.ELECTRON_START_URL ||
    url.format({
      pathname: path.join(__dirname, '/../build/index.html'),
      protocol: 'file:',
      slashes: true,
    });
  mainWindow.loadURL(startUrl);

  mainWindow.on('close', e => {
    if (emitAppWillClose) {
      emitAppWillClose = false;
      mainWindow.webContents.send('app-will-close');
      e.preventDefault();
    }
  });

  // Emitted when the window is closed.
  mainWindow.on('closed', function() {
    // Dereference the window object, usually you would store windows
    // in an array if your app supports multi windows, this is the time
    // when you should delete the corresponding element.
    mainWindow = null;
  });
}

// This method will be called when Electron has finished
// initialization and is ready to create browser windows.
// Some APIs can only be used after this event occurs.
app.on('ready', createWindow);

// Quit when all windows are closed.
app.on('window-all-closed', function() {
  // On OS X it is common for applications and their menu bar
  // to stay active until the user quits explicitly with Cmd + Q
  if (process.platform !== 'darwin') {
    app.quit();
  }
});

app.on('activate', function() {
  // On OS X it's common to re-create a window in the app when the
  // dock icon is clicked and there are no other windows open.
  if (mainWindow === null) {
    createWindow();
  }
});

ipcMain.on('addProcessId', (event, processId) => {
  processIds.push(processId);
});

ipcMain.on('removeProcessId', (event, processId) => {
  processIds = processIds.filter(id => id !== processId);
});

ipcMain.on('killAllRunningProcesses', event => {
  if (processIds.length) {
    event.preventDefault();
    killAllRunningProcesses();
  }
  app.quit();
});

ipcMain.on('triggerClose', (e, proceed) => {
  if (!proceed) {
    // user aborted
    emitAppWillClose = true; // reset flag
    return;
  }

  mainWindow.close();
});

const killAllRunningProcesses = () => {
  try {
    processIds.forEach(processId => {
      killProcessId(processId);

      // Remove the parent or any children PIDs from the list of tracked
      // IDs, since they're killed now.
      processIds = processIds.filter(id => id !== processId);
    });
  } catch (err) {
    console.error('Got error when trying to kill children', err);
  }
};

const canApplicationBeMoved = () => {
  // The application can be moved if :
  //  - The platform is MacOS
  //  - The function 'isInApplicationsFolder' exists
  //  - The app is running in production
  //  - Guppy is not already in the Applications folder
  //  - The user hasn't chosen to not see the dialog prompt again
  const hasApplicationsFolder =
    process.platform === 'darwin' &&
    typeof app.isInApplicationsFolder === 'function';

  // NOTE: Because this file isn't compiled by Webpack, `process.env.NODE_ENV`
  // will be undefined in the bundled application. Rather than check to see if
  // it's set to `production`, we just care that it ISN'T set to `development`
  // (development is set in package.json when running the 'start' script).
  const isProduction = process.env.NODE_ENV !== 'development';

  if (
    hasApplicationsFolder &&
    isProduction &&
    !app.isInApplicationsFolder() &&
    !electronStore.has(MOVE_TO_APP_FOLDER_KEY)
  ) {
    return true;
  }

  return false;
};

const manageApplicationLocation = () => {
  if (canApplicationBeMoved()) {
    dialog.showMessageBox(
      {
        type: 'question',
        buttons: ['Yes, move to Applications', 'No thanks'],
        message: 'Move to Applications folder?',
        detail:
          "I see that I'm not in the Applications folder. I can move myself there if you'd like!",
        icon: path.join(__dirname, icon256Path),
        cancelId: 1,
        defaultId: 0,
        checkboxLabel: 'Do not show this message again',
      },
      (res, checkboxChecked) => {
        // User wants the application to be moved
        if (res === 0) {
          try {
            app.moveToApplicationsFolder();
          } catch (err) {
            dialog.showErrorBox(
              'Error',
              'Could not move Guppy to the Applications folder'
            );
            console.error(
              'Could not move Guppy to the Applications folder',
              err
            );
          }
        }

        // User doesn't want to see the message again
        if (checkboxChecked === true) {
          electronStore.set(MOVE_TO_APP_FOLDER_KEY, true);
        }
      }
    );
  }
};
