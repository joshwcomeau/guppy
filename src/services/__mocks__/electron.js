import * as path from 'path';
module.exports = {
  remote: {
    app: {
      getAppPath: () => path.resolve(__dirname, '..', '..', '..'),
      getPath: () =>
        process.env.APPDATA ||
        (process.platform == 'darwin'
          ? process.env.HOME + 'Library/Preferences'
          : '/var/local'),
    },
  },
};
