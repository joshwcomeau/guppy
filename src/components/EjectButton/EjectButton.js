// @flow
import React, { PureComponent } from 'react';
import IconBase from 'react-icons-kit';
import { ic_eject as ejectIcon } from 'react-icons-kit/md/ic_eject';
import { remote } from 'electron';

import { COLORS } from '../../constants';

import BigClickableButton from '../BigClickableButton';

const { dialog } = remote;

type Props = {
  width: number,
  height: number,
  isRunning: boolean,
  onClick: () => void,
};

class EjectButton extends PureComponent<Props> {
  handleClick = () => {
    dialog.showMessageBox(
      {
        type: 'warning',
        buttons: ['Yes, light this candle', "Ahhh no don't do that"],
        defaultId: 1,
        title: 'Are you sure?',
        message:
          'Ejecting is a permanent one-time task that unwraps the create-react-app environment.',
        detail:
          "It's recommended for users comfortable with Webpack, who need to make tweaks not possible without ejecting.",
      },
      (response: number) => {
        // The response will be the index of the chosen option, from the
        // `buttons` array above.
        const isConfirmed = response === 0;

        if (isConfirmed) {
          this.props.onClick();
        }
      }
    );
  };

  render() {
    const { isRunning } = this.props;

    return (
      <BigClickableButton
        width={40}
        height={34}
        colors={[COLORS.purple[500], COLORS.blue[700]]}
        isOn={isRunning}
        onClick={this.handleClick}
      >
        <IconBase size={24} icon={ejectIcon} />
      </BigClickableButton>
    );
  }
}

export default EjectButton;
