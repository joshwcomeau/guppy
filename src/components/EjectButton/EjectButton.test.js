/* eslint-disable flowtype/require-valid-file-annotation */
import React from 'react';
import { shallow } from 'enzyme';
import { remote } from 'electron'; // Mocked
import EjectButton, { dialogOptions, dialogCallback } from './EjectButton';
import BigClickableButton from '../BigClickableButton';

const { dialog } = remote;

describe('EjectButton component', () => {
  let wrapper;
  let clickHandler;

  beforeEach(() => {
    clickHandler = jest.fn();
    wrapper = shallow(<EjectButton onClick={clickHandler} />);
  });

  it('should render without being pressed (not running)', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render as pressed (running)', () => {
    wrapper.setProps({ isRunning: true });
    expect(wrapper).toMatchSnapshot();
  });

  describe('Confirm dialog', () => {
    let ejectButton;
    beforeEach(() => {
      ejectButton = wrapper.find(BigClickableButton);
      ejectButton.simulate('click');
    });

    it('should display dialog', () => {
      expect(dialog.showMessageBox).toHaveBeenCalledWith(
        dialogOptions,
        expect.anything()
      );
    });

    it('should call clickHandler if confirmed', () => {
      dialogCallback.call(wrapper.instance(), 0);
      expect(clickHandler.mock.calls.length).toBe(1);
    });

    it('should dismiss with-out calling clickHandler', () => {
      dialogCallback.call(wrapper.instance(), 1);
      expect(clickHandler.mock.calls.length).toBe(0);
    });
  });
});
