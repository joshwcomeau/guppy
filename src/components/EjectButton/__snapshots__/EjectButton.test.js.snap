// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EjectButton component should render as pressed (running) 1`] = `
<BigClickableButton
  colors={
    Array [
      "#651fff",
      "#304FFE",
    ]
  }
  height={34}
  isOn={true}
  onClick={[Function]}
  thickness={6}
  width={40}
>
  <Icon
    fill="currentColor"
    icon={
      Object {
        "children": Array [
          Object {
            "attribs": Object {
              "d": "M5 17h14v2H5zm7-12L5.33 15h13.34z",
            },
            "name": "path",
          },
        ],
        "viewBox": "0 0 24 24",
      }
    }
    size={24}
  />
</BigClickableButton>
`;

exports[`EjectButton component should render without being pressed (not running) 1`] = `
<BigClickableButton
  colors={
    Array [
      "#651fff",
      "#304FFE",
    ]
  }
  height={34}
  onClick={[Function]}
  thickness={6}
  width={40}
>
  <Icon
    fill="currentColor"
    icon={
      Object {
        "children": Array [
          Object {
            "attribs": Object {
              "d": "M5 17h14v2H5zm7-12L5.33 15h13.34z",
            },
            "name": "path",
          },
        ],
        "viewBox": "0 0 24 24",
      }
    }
    size={24}
  />
</BigClickableButton>
`;
